proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,2290 @@
+
+That is, in the above example, ``R`` has type
+
+::
+
+  R :: A ->. B ->. R
+
+In general, in
+
+::
+
+  data R where
+    R :: { f1 ::(π) A, f2 ::(ρ) B } -> R
+
+We have
+
+::
+
+  R :: A :π-> B :ρ-> R
+
+With absence of annotation interpreted as annotating with ``'One``.
+
+
+Inference
+---------
+
+.. _Inference:
+
+Because of backwards compatibility, we initially chose the following
+strategy: when the type of a function is not constrained by given
+constraints, we conservatively assume it to have multiplicity ω.
+
+Experience shows that this sometimes yield very confusing error messages
+where perfectly valid code is rejected:
+
+::
+
+  type family L x
+  type instance L Int = A ->. A
+
+  f :: L x -> x
+
+  u :: Int
+  u = f (\x -> x)
+
+While the identity function is indeed linear, because the resolution
+of the type family (``L Int ~ Int``) is delayed in GHC, ``\x -> x`` is
+considered to have no given type, and is inferred to have a non-linear
+type, and is refused by the type-checker.
+
+We therefore need a more refined strategy, to avoid surprising
+behaviour like the above. We do not expect it to be too hard to
+implement a better strategy, but we don't have a specification yet.
+
+A more profound difficulty exists for inference: for explicit ``let``
+bindings and ``case`` expressions (*i.e.* which are not generated from
+the desugaring of an equation but are written as ``let``, ``where``,
+or ``case`` in the surface syntax), we want to infer the multiplicity
+annotation. The process for this is not yet defined (see `Unresolved
+questions`_ below for a more precise description of this issue).
+
+Non-termination, exceptions & catch
+-----------------------------------
+
+.. _Exceptions:
+
+In the presence of non-termination or exceptions, linear functions may
+fail to fully consume their argument. We can think of it as: the
+consumption of the result of the function was never complete, so the
+consumption of the argument need not be either. However, because
+exceptions can be caught, a program can observe a state where a value
+``v`` has been passed to a linear function ``f`` but the call ``f v``
+has exited (with an exception) without consuming ``v``. So while, the
+guarantee provided by linear functions holds for converging
+computations, we must weaken it in case of divergence:
+
+- Attempting to consume exactly once ``f v``, when ``f`` is a linear
+  function, will consume ``v`` exactly once if the consumption of ``f
+  v`` converges, and *at most once* if it diverges.
+
+Where "consuming at most once" is defined by induction, like
+"consuming exactly once", but every sub-consumption is optional.
+
+In the paper, we gave a simplified specification of a linear ``IO``
+monad (called ``IOL``) which ignored the issue of exception for the
+sake of simplicity. Can we, still, write a resource-safe ``RIO`` monad
+with linear types despite the added difficulty of exceptions? Yes, as
+this section will show.
```

Comment:
Opening a subthread here to answer @simonmar 's concerns.

>Most resources are allocated with a bracket pattern (`bracket allocate deallocate $ \resource -> do ...`) or using a `withFoo combinator`. These give clear scoping and exception-safety to resource allocation, and linear types don't offer any safety improvements here - as I understand it the bracket pattern will remain the safest way to manage explicit resources in cases where it can be used.

The goal of this section, currently §2.14 (Non-termination, exceptions & catch), and of the [corresponding library](https://github.com/tweag/linear-base/blob/master/src/System/IO/Resource.hs) is to demonstrate that, actually, it's quite possible to build a linear abstraction which will be safer than brackets.

> The ones that spring to mind are where resources are allocated in C++ and released in Haskell or allocated in Haskell and then released in a different Haskell thread after receiving a message from C++.

It sounds like a similar use-case as [this blog post](https://www.tweag.io/posts/2017-11-29-linear-jvm.html).

> our experience has been that safety in these cases has only been an issue where exceptions are concerned - and linear types doesn't help with that

Does section §2.14 convince you that, in fact, linear types can help you build abstractions where the program logic doesn't have to deal with exceptions, while ensuring proper release?

Reply:
> Does section §2.14 convince you that, in fact, linear types can help you build abstractions where the program logic doesn't have to deal with exceptions, while ensuring proper release?

Actually no. I think `RIO` brings different problems. Of course there are the obvious drawbacks that it adds overhead, requires wrapping all your resource allocation and deallocation functions, and adds another monad layer which is inconvenient. But aside from these, I think exception handling is problematic.

Would you allow exceptions to be handled inside RIO?

* If yes, then we lose the guarantee of prompt resource release in the event of an exception. Resources would be released later than if we had used bracket, perhaps arbitrarily later.
* If no, then we have to drop back to IO to do exception handling, and we can't have resources whose lifetimes overlap with an exception-handling operation.

Reply:
@simonmar let's try and take this one step at a time. Your concerns, as I understood them, were that, were you to use linear types for resource management, you would sill need to use brackets in your code (thus negating the benefits on linear types). Do you agree that `RIO` actually alleviates the need for bracket? If so, let's discuss catch; if not, let's discuss brackets further.

Reply:
> Your concerns, as I understood them, were that, were you to use linear types for resource management, you would sill need to use brackets in your code (thus negating the benefits on linear types)

Correct. But to be clear - I don't think there's anything wrong with `bracket`, and I'm not looking for a better solution.  My concern was that linear types don't actually buy anything for this kind of resource management, which seems like a very common case at least from my perspective.

Could I replace `bracket allocate deallocate $ \resource -> X` directly with `runRIO $ do resource <- allocate; liftIO (X resource); deallocate resource`? Sure - but I don't think this gains very much, I've just replaced `bracket` with `runRIO`. It stops you from returning the resource and using it outside the `bracket`, but if I really want that I could also use the `runST` trick. Bugs of this kind are not a big problem. Perhaps I get to deallocate the resource a little bit earlier than with `bracket`, but that's a very marginal gain, I've never encountered a case where I needed more promptness than `bracket` provided.

Reply:
I would suggest that explicit (checked) resource management is less about promptness, which is nice, than _interleaving_.  `RIO` lets you acquire resources and release them _in any causal order_. (i.e. you may allocate X, do stuff, allocate Y, deallocate X, do stuff that _requires_ Y allocated but X deallocated, and finally deallocate Y). For filehandles this is typically not the case, but for e.g. database or thread locks, etc, it is very imaginable. The `runST` trick doesn't get you that either, unless you add in a whole other style of explicit type-level hacking for commuting its existential witnesses.

(I'd also add that accidentally returning resources outside of a bracket-like thing has bit me and other people I consider skilled more often than i'd like to admit, in tricky situations. I spent lot of time the other week trying to understand why something failed before realizing that it was such a situation in deep disguise (there were some iteratees in the way). By the way, the other trick there turned out to be `Codensity` which let different local CPS transforms effectively merge scope. `RIO` may well have helped make that subtle problem more obvious).

Reply:
I don't dispute that there are situations where `bracket` doesn't fit, my claim was that they're in a small minority.

@gbaz In the example you mention with interatees and `Codensity`, would it be straightforward to manage resources with linear types? How much would you need to change the other abstraction layers being used here?

Reply:
Oh everything would have to be rewritten into a totally different style of resource management (using explicit allocs/deallocs with affine discipline). But if we _did_ use this paradigm pervasively, I could see how it would benefit such situations!

Reply:
> Could I replace `bracket allocate deallocate $ \resource -> X` directly with `runRIO $ do resource <- allocate; liftIO (X resource); deallocate resource`?

I just wanted to add that this is really not what is advertised. I really don't see a point in doing that (sure, there is extra safety, but clearly not worth the cost. Instead, the proposal is to replace  `bracket allocate deallocate $ \resource -> X` with

```haskell
  resource <- allocate
  X
  deallocate resource
```

Inside of a larger `runRIO` block.

Reply:
Right, and that brings us back to my earlier point - if we have larger `runRIO` blocks, then at some point we encounter a situation where we need to do exception handling. Should exception handling be done in the `RIO` monad or not?  (see https://github.com/ghc-proposals/ghc-proposals/pull/111#discussion_r201929958)

I think the answer *must* be "no", because `RIO` is providing a guarantee of prompt resource deallocation in the event of an exception, and to allow exception handling inside `RIO` itself would lose that guarantee, and we'd be in a worse place than if we'd used `bracket`. But if I can't do exception handling in `RIO`, then that means every time I need to do exception handling I have to drop back to `IO` and start a new `runRIO`, and resources can't have lifetimes that span an exception-handling point.

Reply:
Sorry for not rounding off this thread earlier. I think we agree on the details, so let me take a step back and restate the concern that started this.

A big part of the motivation for the proposal is safe resource handling. But (at least to my mind) doing resource handling using linear types comes with some significant drawbacks, and in some cases might not be possible at all. To summarize the drawbacks: (a) you have to use a finalizer or some mechanism to release the resources in the event of an exception, such as the `RIO` monad, and (b) you cannot use linear resources in conjunction with exception handling in the `IO` monad. I'd really like it if the proposal could mention this, because I think many potential uses of linear types would bump into it.

Reply:
Yes, but remember: 

* Exceptions mean that the "used exactly once" guarantee may be broken
* But for some purposes that may not matter.  E.g. If I'm sent a channel in a message from a server, and I'm supposed to reply on it, then my entire computer might die; so the server must eal with the no-reply situation.  So, failing to reply if there's a divide-by-zero, say, is perfectly acceptable.
* Meanwhile, for the non-exceptional path it is a Huge Win to have static checking that I do not reply twice, and do not simply discard the channel.  That's a real benefit.
* Yes, if you absolutely need finalization, you'll still need to arrange that, with a finalizer or RIO or whatever.  But even then the static checks for the non-exceptional path are extremely valuable.  (And finalizers are not prompt, so this also guarantees prompt finalization for the non-exceptional path.)

Granted, the proposal could be clearer about this.  But I don't think we should let the best (guaranteed finalization even in the presence of arbitrary exceptions -- unachievable) should be the enemy of the (highly achievable) good.

Reply:
@simonmar I devote the section _Non-termination, exceptions & catch_ (currently 2.14) to this discussion. Can you let me know what you think should be improved in this section? It would be tremendously useful!

Reply:
> But this is an absolute requirement for any reliable system!

Not in all cases.  If the entire machine dies, no amount of bracketing is going to send a response back to the server!    

> Don't virtually all resources need to be reliably released?

Well, Erlang is architected around the idea of "just kill the process" if anything goes wrong, with no attempt at recovery or remediation on any finer grain than that.

And finalizers do a reasonable job of eventual release; they just aren't prompt.

> In the majority of cases, releasing a resource is already handled well by bracket and withFoo abstractions

Again, not for the "send a reply on this channel" example.  I suppose you could put a flag in the channel to say if the reply has been sent; the exception hander looks at the flag, sends ABORT if the reply has not been sent.   

We aren't disagreeing here really!

Reply:
We aren't disagreeing, but I think we're talking about slightly different things. By a "resource" I mean some object that must be explicitly released. Of course if the machine dies or the process is killed then the resource will be recovered, either by the OS in the case of the process or by rebooting in the case of the machines dying, or whatever. But if the program continues running then it is an absolute requirement that the program explicitly releases the resource, unless you're happy for your processes to run out of memory and die every so often. If linear resources came with some unavoidable leaks, that would be bad, right? Because you wouldn't be able to build long-running programs.

Yes there are things that *don't* need to be actively released. I'm not talking about those.

Even in Erlang you can't get away without deallocating your resources. I googled a bit and came across this: http://erladvisor.blogspot.com/2015/03/resource-management-idioms-in-erlang.html which looks like the basic idioms are a kind of finalization which looks similar to RIO, and exception handling which is similar to bracket.

Once again let me try to state the problem taking into account discussion so far:
* Most explicitly allocated resources need explicit deallocation.
* We have the option of using `bracket`/`withFoo` or linear types to ensure safe deallocation. 
* Linear types requires (a) RIO or finalizers, neither of which is very satisfactory, and (b) can't be used if you need to handle exceptions in the IO monad.

`bracket`/`withFoo` also has some unsafety - nothing checks that the resource doesn't escape its scope. But on the whole I think most explicit resource deallocation will be better off with the `bracket/withFoo` idiom. That's not to say that it won't be a good idea to use linear types for other properties that don't need an "exactly once" guarantee (e.g. "send a reply on this channel").